backend-worker/ucm_valuation.py

```diff
@@ -0,0 +1,494 @@
+            lower_threshold = mortality_risk_df.loc[0, f"t_{t}"]
+            upper_threshold = mortality_risk_df.loc[0, f"t_{thresholds[i]}"]
+
+            # Mortality Risks
+            lower_risk = mortality_risk_df.loc[0, f"rr_{t}"]
+            upper_risk = mortality_risk_df.loc[0, f"rr_{thresholds[i]}"]
+
+            # Calculate mask
+            # Initial bin
+            if i == 1:
+                current_mask = np.logical_and(
+                    valid_mask, (t_air_array < upper_threshold)
+                )
+            # Final Bin
+            elif i == len(thresholds) - 1:
+                current_mask = np.logical_and(
+                    valid_mask, (t_air_array >= lower_threshold)
+                )
+            # All other bins
+            else:
+                current_mask = np.all(
+                    (
+                        valid_mask,
+                        (t_air_array >= lower_threshold),
+                        (t_air_array < upper_threshold),
+                    ),
+                    axis=0,
+                )
+
+            # Actual calculation
+            t_air_masked = t_air_array[current_mask]
+            mortality[current_mask] = (t_air_masked - upper_threshold) / (
+                lower_threshold - upper_threshold
+            ) * (lower_risk - upper_risk) + upper_risk
+
+        return mortality
+
+    pygeoprocessing.raster_calculator(
+        [(str(t_air_raster_path), 1)],
+        mortality_op,
+        str(target_mortality_path),
+        gdal.GDT_Float32,
+        TARGET_NODATA,
+    )
+
+
+def grouped_scalar_calculation(
+    base_raster_path: pathlike,
+    category_raster_path: pathlike,
+    target_raster_path: pathlike,
+    category_list: List[int],
+    scalar_list: List[int],
+) -> None:
+    """Raster calculator that multiplies some base raster by scalars associated with a
+    different raster's categories
+
+    Args:
+        base_raster_path (pathlike): Path to base raster to multiply
+        category_raster_path (pathlike): Path to raster with categories
+        target_raster_path (pathlike): Path to target raster
+        category_list (List[int]): List of categories in category raster
+        scalar_list (List[int]): List of scalars associated with categories
+
+    Returns:
+        None
+
+    """
+    # Ensure path variables are Path objects
+    base_raster_path = Path(base_raster_path)
+    category_raster_path = Path(category_raster_path)
+    target_raster_path = Path(target_raster_path)
+
+    temporary_working_dir = Path(tempfile.mkdtemp(dir=target_raster_path.parent))
+
+    base_raster_info = pygeoprocessing.get_raster_info(str(base_raster_path))
+    base_raster_nodata = base_raster_info["nodata"][0]
+    cell_size = np.min(np.abs(base_raster_info["pixel_size"]))
+
+    category_raster_nodata = pygeoprocessing.get_raster_info(str(category_raster_path))[
+        "nodata"
+    ][0]
+
+    # Calculate kWh map
+    grouped_scalar_op = MultiplyRasterByScalarList(
```

Comment:
Using a class here is a clever solution for doing some more complicated `raster_calculator` operations. There are a couple other ideas that come to mind too.

A. additional data can be passed through to the `raster_calculator` `op` alongside the input rasters, by adding tuples like `(scalar_list, "raw")` to the first argument. I think that would allow the `op` to be a plain function instead of a class.

or 

B. If I'm understanding the objective correctly, I think I see two separate operations:
1. `reclassify` an LULC raster such that the categorical pixel values are replaced by their corresponding scalar value
2. `raster_calculator` to multiply the base raster (GDD, etc) by the new raster of scalars.
Does that seem like it accomplishes the same?

I'm not suggesting you should refactor things to either of these solutions. I just wanted to mention some alternatives since using classes often adds some complexity & verbosity.

Reply:
I think both of these options accomplish what I'm doing here with the class. I'm totally fine with your option A, but option B adds additional intermediate files that I would rather not create.

In general, I've started using these sorts of class wrappers for pygeoprocessing.raster_calculator ops in a lot of my code. While I still find classes to be a confusing structure in python, I appreciate the ability to pre-load a function with a nodata value or with other variables such as these scalar lists. Just a personal preference.

Reply:
To deal with the verbosity I've sequestered all of these sorts of class rasterops into a module within my own personal python package that I call for project applications. Here I've copied in the class definition to untangle this version from my own coding workspace.